pl_env.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2021 Patrick Gaskin
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "pl_env.h"
+#include "options.h"
+#include "xmalloc.h"
+
+#include <ctype.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+
+extern char **environ;
+static char **pl_env_cache = NULL;
+
+static bool pl_env_contains_delimiter(const char *str)
+{
+	return !!strchr(str, PL_ENV_DELIMITER);
+}
+
+/**
+ * pl_env_norm normalizes the provided path into the required format for
+ * pl_env_save and pl_env_restore, which does exact string matching/replacement.
+ * It converts backslashes to slashes on Windows, removes consecutive slashes,
+ * removes './' path segments, simplifies '../' path segments (and returns NULL
+ * if it would result in going above the uppermost directory), and removes the
+ * trailing slash.
+ */
+static char *pl_env_norm(char *path)
+{
+#ifdef _WIN32
+	/* convert backslashes to slashes */
+	/* note: cmus uses forward slashes internally, but Windows accepts both */
+	for (char *p = new; *p; p++)
```

Comment:
I'll fix the variable. It got lost during refactoring since it was behind an ifdef.

> By the way does WSL / Cygwin / whatever not give us a unix-style view of the filesystem?

The issue is if the user passes a path. The compatibility layers generally do accept both, but for `pl_env` it needs to match exactly, so we normalize it.

Reply:
Fixed.